### PR TITLE
Paginate data list responses after a configurable threshold 

### LINKED ISSUE #2008: Data ViewSet: Automatically paginate requests after a certain submission threshold
## Suggested Feature / Enhancement

Currently, the data viewset returns all the instances present in a form if the `page_size` query param is not specified. It would be nice if the response was paginated if a form has a lot of submissions (10k+ perhaps ?) in order to reduce the amount of processing it may take to retrieve all the records.

### PR DESCRIPTION
**Pull request by DavisRayM**
### Changes / Features implemented

- Paginate responses when requested data is more than the retrieval threshold

### Steps taken to verify this change does what is intended

- Added tests

### Side effects of implementing this change

Closes #2008

### CODE REVIEW

**Review by ivermac — APPROVED** [review]
Looks good to me

File: onadata/apps/api/viewsets/data_viewset.py [hunk 1]
```diff
@@ -617,17 +617,29 @@ def _get_data(self, query, fields, sort, start, limit, is_public_request):
...
             query, fields, sort, start, limit, is_public_request)
 
+        retrieval_threshold = getattr(
+            settings, "SUBMISSION_RETRIEVAL_THRESHOLD", 10000)
         pagination_keys = [self.paginator.page_query_param,
                            self.paginator.page_size_query_param]
         query_param_keys = self.request.query_params
         should_paginate = any([k in query_param_keys for k in pagination_keys])
+
+        if not should_paginate and not is_public_request:
```
Comment (ukanga) [inline onadata/apps/api/viewsets/data_viewset.py:627]: Does this mean we will not paginate public forms that exceed the threshold?

Comment (DavisRayM) [reply onadata/apps/api/viewsets/data_viewset.py:627]: At the moment yes; didn't think of this while implementing the feature... Any thoughts/reservations with always paginating public requests? _Seems a bit painful to calculate the entire count(getting all public XForms and summing the `num_of_submissions` of public data that's currently available..._

Comment (ukanga) [reply onadata/apps/api/viewsets/data_viewset.py:627]: This is for a single form request at a time, not all public forms at the same. Also, we could go with `num_of_submissions` that is not calculated on the fly, it is a safe estimate in most situations.

Comment (ukanga) [reply onadata/apps/api/viewsets/data_viewset.py:627]: I withdraw my statement, I see that is_public_request represents access to the `/api/v1/data/public` and not to one single form data.

**Review by ukanga — APPROVED** [review]

**Pull request merged**